### ISSUE TITLE
Network policy is blocking all network traffic

### ISSUE THREAD

**Issue by asher-lab** [cilium/cilium#28667]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

After applying the following netpol, I can't reach out to any IP except for the IP of the pod itself. 
```
root@cks1:~/exam# kg netpol -A
NAMESPACE   NAME                               POD-SELECTOR   AGE
default     allow-all   <none>         5m3s



apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all
spec:
  podSelector: {}
  policyTypes:
    - Egress
    - Ingress
  egress:
    - to:
        - ipBlock:
            cidr: 0.0.0.0/0
  ingress:
    - from:
        - ipBlock:
            cidr: 0.0.0.0/0
```

Without the netpol, of course I can reach out to other website or IP.

```
kg netpol -A
No resources found
root@cks1:~/exam# kg pods -owide
NAME         READY   STATUS    RESTARTS   AGE   IP           NODE   NOMINATED NODE   READINESS GATES
be           1/1     Running   0          10m   10.0.0.43    cks1   <none>           <none>
fe           1/1     Running   0          10m   10.0.0.63    cks1   <none>           <none>
simple-app   1/1     Running   0          10m   10.0.0.149   cks1   <none>           <none>
root@cks1:~/exam# k exec fe -- curl 1.1.1.1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   167  100   167    0     0  32370      0 --:--:-- --:--:-- --:--:-- 33400
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>cloudflare</center>
</body>
</html>

```

I have tried to use another CNI which is Calico and with the same netpol, I can reach out to other IP or website without any problem. This might be a bug??

- Btw I have bootstrapped my cluster already using this guide both w. cilium and w.o cilium proxy as well https://docs.cilium.io/en/stable/installation/k8s-install-kubeadm/ 
- I also tried using Digital Ocean VM aside from Azure and still get the same results

Thank you team.

### Cilium Version

Client: 1.14.2 a6748946 2023-09-09T20:59:33+00:00 go version go1.20.8 linux/amd64
Daemon: 1.14.2 a6748946 2023-09-09T20:59:33+00:00 go version go1.20.8 linux/amd64

### Kernel Version

Linux cks1 5.15.0-1049-azure #56~20.04.1-Ubuntu SMP Thu Sep 21 13:01:28 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.2

### Sysdump

```
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
Error: unknown command "sysdump" for "cilium"
Run 'cilium --help' for usage.
unknown command "sysdump" for "cilium"
command terminated with exit code 1
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by joestringer** [comment]
Hello, thanks for the report. Are you attempting to reach other Pods within the cluster or IPs outside the cluster? There is a known behaviour with Cilium's interpretation of CIDR block policies that means they will only apply to IPs that are outside the cluster: https://docs.cilium.io/en/stable/network/kubernetes/policy/#networkpolicy-state . If you are attempting to connect to Pods inside the cluster, I suggest using label selectors.

**Comment by asher-lab** [comment]
Hello @joestringer thanks for the reply. I am trying to connect from IP's outside the cluster. However when applying that network policy, I am not able to reach anything outside the cluster (e.g. curling google.com). From what I understand, that network policy above should allow all outside traffic from reaching the cluster.

Below is an example:
```
 k exec fe -- curl google.com
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:19 --:--:--     0curl: (6) Could not resolve host: google.com
command terminated with exit code 6
```
However when doing the IP it allows me to reach google.com:
```
root@cks1:~/cilium# k exec fe -- curl google.com
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:19 --:--:--     0curl: (6) Could not resolve host: google.com
command terminated with exit code 6
root@cks1:~/cilium# ^C
root@cks1:~/cilium# k exec fe -- curl 172.217.194.102
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   219  100   219    0     0   1443      0 --:--:-- --:--:<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="http://www.google.com/">here</A>.
</BODY></HTML>
-- --:--:--  1450
```

**Comment by joestringer** [comment]
I would suggest using Hubble to observe the traffic flows and understand what is happening to the packets.

**Comment by MaxAnderson95** [comment]
I too am having this issue. Initially experienced using 1.14 on RKE2 1.24, but I can reproduce it in Kind on Linux as well:

1. Create a kind config
```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
networking:
  disableDefaultCNI: true   # do not install kindnet
  kubeProxyMode: none       # do not run kube-proxy
nodes:
- role: control-plane
- role: worker
- role: worker
- role: worker
```

2. Create a cluster `kind create cluster --config config.yaml`
3. Create a Cilium Helm values file
```yaml
kubeProxyReplacement: strict
k8sServiceHost: kind-control-plane # use master node in kind network
k8sServicePort: 6443               # use api server port
hostServices:
  enabled: false
externalIPs:
  enabled: true
nodePort:
  enabled: true
hostPort:
  enabled: true
image:
  pullPolicy: IfNotPresent
ipam:
  mode: kubernetes
hubble:
  enabled: true
  relay:
    enabled: true
```
4. Install Cilium `helm upgrade --install --namespace kube-system --repo https://helm.cilium.io cilium cilium --values values.yaml`
5. Create a namespace and pod that continuously curls Google
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: v1
kind: Pod
metadata:
  name: curl-pod
  namespace: test
spec:
  containers:
  - name: curl-container
    image: curlimages/curl:latest
    command: ["/bin/sh", "-c"]
    args:
    - |
      while true; do
        curl --max-time 1 https://www.google.com
        sleep 1
      done
  restartPolicy: Never
```
6. Inspect the logs and see the pod can communicate `kubectl logs curl-pod -n test -f`
7. Apply the following Egress NetworkPolicy that should allow this communication to continue
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all
  namespace: test
spec:
  podSelector: {}
  policyTypes:
    - Egress
  egress:
    - to:
        - ipBlock:
            cidr: 0.0.0.0/0
```
6. Inspect the logs again and see the pod is being blocked and the requests are timing out `kubectl logs curl-pod -n test -f`

If I do these steps while using Calico/Canal it works as expected. This occurs on RKE2 and Kind, with an without kube-proxy replacement mode.

**Comment by MaxAnderson95** [comment]
Adding some additional information....

- I've tried this with DigitalOcean Kubernetes which uses Cilium and get the same result. Three different distributions of Kubernetes, so this indicates a Cilium problem most likely.
- Interestingly, I get the same result (traffic blocked unexpectedly) when using the `CiliumNetworkPolicy` CDR and crafting an equivalent policy:
```yaml
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: allow-all
  namespace: test
spec:
  endpointSelector:
    matchLabels: {}
  egress:
    - toCIDR: ["0.0.0.0/0"]
```

**Comment by squeed** [comment]
@asher-lab

> I am trying to connect from IP's outside the cluster. However when applying that network policy, I am not able to reach anything outside the cluster (e.g. curling google.com). 

That's exactly what that policy is doing :-). Specifically, since pods with this policy cannot communicate with the cluster, they cannot access the kube-dns server within the cluster. That's why you're not able to resolve the **name** for `google.com`, even if you can communicate with the IP address.

I suggest you modify your policy to use the [`all` entity](https://docs.cilium.io/en/latest/security/policy/language/#entities-based)

**Comment by MaxAnderson95** [comment]
I can't speak for @asher-lab but I have tested this extensively and it's not a DNS issue (though you're right the rule needs to allow DNS traffic). If I spin up a testing pod and exec into it, I can't ping public addresses by IP (ex. 8.8.8.8) despite being allowed by the network policy with 0.0.0.0/0.

I can post some additional steps to reproduce when I get back to a computer, but basically, I discovered this issue when installing a 3rd party chart onto a cluster using Cilium, and some of the pods that needed Kube API access were crashing because they couldn't get access, despite the chart also including appropriate network policies. Installing the same chart onto an identical cluster but with Antrea instead of Cilium solved the issues.
